MINOR: Fix Streams EOS system tests by adding clean-up of state dir

cadonna opened the pull request:
Recently, system tests `test_rebalance_[simple|complex]` failed
repeatedly with a verfication error. The cause was most probably
the missing clean-up of a state directory of one of the processors.

A node is cleaned up when a service on that node is started and when 
a test is torn down.

If the clean-up flag `clean_node_enabled` of a EOS Streams service is 
unset, the clean-up of the node is skipped.

The clean-up flag of `processor1` in the EOS tests should stay set before
its first start, so that the node is cleaned before the service is started.
Afterwards for the multiple restarts of `processor1` the cleans-up flag should
be unset to re-use the local state.

After the multiple restarts are done, the clean-up flag of processor1 should
again be set to trigger node clean-up during the test teardown.

A dirty node can lead to test failures when tests from Streams EOS tests are
scheduled on the same node, because the state store would not start empty
since it reads the local state that was not cleaned up.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cadonna committed:
HOTFIX: Fix Streams EOS system tests by adding clean-up of state dir

Recently, system tests `test_rebalance_[simple|complex]` failed
repeatedly with a verfication error. The cause was most probably
the missing clean-up of a state directory of one of the processors,
which led to an unexpected state and in turn to unexpected results
in the output topic.

This commit adds the missing clean-up to the EOS system tests.

The clean-up was also not present in system tests
`test_failure_and_recovery*`. Although those system have not failed
recently, I also added the fix there.

cadonna commented on tests/kafkatest/tests/streams/streams_eos_test.py:
```diff
@@ -67,6 +67,9 @@ def run_rebalance(self, processor1, processor2, processor3, verifier):
         self.stop_streams2(processor1, processor3)
+
+        processor1.clean_node_enabled = True
+
```
On line 61, this flag is set to `False` to avoid restoring the state from scratch from the changelog topic for `processor1` when `processor1` is stopped on line 66. However, when `processor1` is stopped on line 73 the state directory needs to be cleaned up because the test ends.

cadonna commented on tests/kafkatest/tests/streams/streams_eos_test.py:
```diff
@@ -109,6 +112,9 @@ def run_failure_and_recovery(self, processor1, processor2, processor3, verifier)
         self.stop_streams2(processor1, processor3)
+
+        processor1.clean_node_enabled = True
+
```
See my comment above. Similar applies here.

cadonna commented:
I am still wondering why `test_rebalance_simple` and `test_rebalance_complex` have always failed together and why `test_failure_and_recovery` and `test_failure_and_recovery_complex` have never failed recently, although they miss the same state directory clean-up.

Locally, the system tests with this fix have not failed so far.

I triggered the system tests with this fix here:
 https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3521/

cadonna commented:
Call for review: @mjsax @bbejeck @ableegoldman

bbejeck commented:
The previous system test run (10 repeats) passed.  I kicked off the same test again https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3522/

bbejeck reviewed (approved):
The changes LGTM, but I'm thinking did this test fail intermittently before or has it just started?

bbejeck commented:
Second round of system tests passed http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-15--001.1573832075--cadonna--fix_test_rebalance--352dab1/report.html

mjsax reviewed (approved):
LGTM.

mjsax commented:
Can we merge this? @cadonna what branches should this be cherry-picked to?

cadonna commented:
@bbejeck, the test started to fail around the November 9th. This PR fixes the local runs of the system tests, but I need some more time to solve it completely.

cadonna commented:
@mjsax Into `trunk` and `2.4`. However, I guess it is not needed on 2.4 before the release..

cadonna commented:
Please wait with merging this PR. I would still like to change it.

cadonna committed:
Move setting of clean-up flag

A node is cleaned up when a service on that node is started and
when a test is torn down.

If the clean-up flag of a EOS Streams service is unset, the clean-up
of the node is skipped.

The clean-up flag of processor1 in the EOS tests should stay set before
its first start, so that the node is cleaned before the service is started.
Afterwards for the multiple restarts of processor1 the cleans-up flag should
be unset to re-use the local state.

After the multiple restarts are done, the clean-up flag of processor1 should
again be set to trigger node clean-up during the test teardown.

A dirty node can lead to test failures when tests from Streams EOS tests are
scheduled on the same node, because the state store would not start empty
since it reads the local state that was not clean up.

cadonna commented on tests/kafkatest/tests/streams/streams_eos_test.py:
```diff
@@ -58,16 +58,16 @@ def run_rebalance(self, processor1, processor2, processor3, verifier):
-        processor1.clean_node_enabled = False
-
         self.add_streams(processor1)
+        processor1.clean_node_enabled = False
```
We can move the unsetting of the clean-up flag after the first start of `processor1` to ensure that the node is cleaned up during the start (see [code line in ducktape](https://github.com/confluentinc/ducktape/blob/f540a740222ca6eaf703d7315228e67b65db3aa9/ducktape/services/service.py#L228)).

cadonna commented on tests/kafkatest/tests/streams/streams_eos_test.py:
```diff
@@ -58,16 +58,16 @@ def run_rebalance(self, processor1, processor2, processor3, verifier):
         self.stop_streams3(processor1, processor3, processor2)
         self.stop_streams2(processor1, processor3)
         self.stop_streams(processor1)
+        processor1.clean_node_enabled = True
```
We can set the clean-up flag here again to ensure that the node is cleaned when the test is torn down.

cadonna commented on tests/kafkatest/tests/streams/streams_eos_test.py:
```diff
@@ -110,6 +109,7 @@ def run_failure_and_recovery(self, processor1, processor2, processor3, verifier)
         self.stop_streams2(processor1, processor3)
         self.stop_streams(processor1)
+        processor1.clean_node_enabled = True
 
```
We do similar here because that was actually the issue in the system tests. Tests `test_failure_and_recovery()` and  `test_failure_and_recovery_complex()` were executed before `test_rebalance_simple` and `test_rebalance_complex` and they were scheduled on the same worker node, respectively. Thus, the latter tests read the local state of the former tests instead of starting with an empty state.

cadonna reviewed (commented):
I had another look at the system test failures and now I am pretty sure that this PR fixes the issues that we have there. I observed that `test_rebalance_simple` and `test_rebalance_complex` were scheduled on the same nodes as `test_failure_and_recovery` and `test_failure_and_recovery_complex` in the failure cases but not in the good cases.

The question arises why only the local store stayed on the nodes and not also the log files of the previous test runs. Locally, I could observe that log files of test runs were concatenated because of the missing clean-up but in the remote system tests I could not. The reason is that within the remote system tests log files are compressed and copied to a shared folder. After compression the logs are removed as can be seen from this log line:
```
[DEBUG - 2019-11-17 07:29:16,673 - remoteaccount - _log - lineno:160]: ubuntu@worker17: Running ssh command: cd "$(dirname /mnt/streams/streams.log)" && f="$(basename /mnt/streams/streams.log)" && if [ -e "$f" ]; then tar czf "$f.tgz" "$f"; fi && rm -rf /mnt/streams/streams.log
``` 
Why did the failures only happen recently?
My theory is that the parallelization of the nightly system tests surfaced the bug. Probably, the parallelization increased the probability that the EOS tests were scheduled to the same nodes.

Note, that it does not make too much sense to run system tests with the branch builder job on this PR, because the tests are run sequentially and would probably not surface the bug. For a general sanity check I will run them, though.

cadonna commented:
The system tests job: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3552/

omkreddy commented:
@bbejeck @cadonna Can we cherry-pick this to 2.4? We are getting similar system test failures in 2.4 branch.

bbejeck commented:
@omkreddy yes, doing so now.

bbejeck commented:
cherry-picked to 2.4

vvcephei commented:
I'm seeing some failures on older branches, so I'm cherry-picking back.

Cherry-picked to 2.3. Confirmed the test worked locally:
```
[INFO:2020-06-01 08:43:27,309]: RunnerClient: kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex: PASS
```

vvcephei commented:
Cherry picked to 2.2:
```
================================================================================
test_id:    kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex
status:     PASS
run time:   4 minutes 37.793 seconds
--------------------------------------------------------------------------------
```

vvcephei commented:
Cherry-picked to 2.1:
```
test_id:    kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex
status:     PASS
run time:   3 minutes 41.565 seconds
```

(also cherry-picked #6898 to enable running the system tests locally)

vvcephei commented:
Cherry-picked to 2.0 (along with #6898)

```
test_id:    kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex
status:     PASS
run time:   3 minutes 33.028 seconds
```

vvcephei commented:
Cherry-picked to 1.1 (along with #6797 and #6898 so that it's possible to actually run the test in docker)

```
test_id:    kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex
status:     PASS
run time:   3 minutes 36.853 seconds
```

vvcephei commented:
Cherry-picked to 1.0 (along with #6797 and #6898)
```
test_id:    kafkatest.tests.streams.streams_eos_test.StreamsEosTest.test_rebalance_complex
status:     PASS
run time:   4 minutes 6.105 seconds
```
